eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/LedgerSpec.hs

```diff
@@ -212,9 +212,30 @@ spec = do
         mkBasicTxBody & withdrawalsTxBodyL .~ Withdrawals [(ra, mempty)]
 
   describe "Mempool" $ do
+    let
+      submitFailingMempoolTx cause tx expectedFailures = do
+        globals <- use impGlobalsL
+        nes <- use impNESL
+        slotNo <- use impLastTickG
+        let
+          mempoolEnv = mkMempoolEnv nes slotNo
+          ls = nes ^. nesEsL . esLStateL
+        case applyTx globals mempoolEnv ls tx of
+          Left err ->
+            err `shouldBe` ApplyTxError @era (injectFailure <$> expectedFailures)
+          Right _ -> assertFailure $ "Expected failure due to " <> cause <> ": " <> show tx
+
+    it "Duplicate transactions" $ do
+      inputs <- replicateM 10 $ do
+        addr <- freshKeyAddr_
+        amount <- Coin <$> choose (2_000_000, 8_000_000)
+        sendCoinTo addr amount
+      tx <- submitTx $ mkBasicTx $ mkBasicTxBody & inputsTxBodyL <>~ fromList inputs
+      submitFailingMempoolTx "duplicate transaction" tx $
+        pure . ConwayMempoolFailure $
+          "All inputs are spent. Transaction has probably already been included"
```

Comment:
Can you also add here a test that submits this transaction:
```haskell
input <- elements inputs
submitFailingTx $ mkBasicTx $ mkBasicTxBody & inputsTxBodyL <>~ Set.singleton input
```

Reply:
I've added that, although another approach would be to make the number of inputs in the first test `choose (1, 10)` or `choose (1, 5)`.

Reply:
That's fine. One is enough to prove the correctness